Code/Framework/AzCore/AzCore/std/containers/span.h

```diff
@@ -0,0 +1,159 @@
+        span(AZStd::array<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        span(AZStd::vector<value_type>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        template<AZStd::size_t N>
+        span(AZStd::fixed_vector<value_type, N>& data)
+            : m_begin(data.data())
+            , m_end(m_begin + data.size())
+        { }
+
+        span(const span&) = default;
+
+        span(span&& other)
+            : span(other.m_begin, other.m_end)
+        {
+#if AZ_DEBUG_BUILD // Clearing the original pointers isn't necessary, but is good for debugging
+            other.m_begin = nullptr;
+            other.m_end = nullptr;
+#endif
+        }
+
+        span& operator=(const span& other) = default;
+
+        span& operator=(span&& other)
+        {
+            m_begin = other.m_begin;
+            m_end = other.m_end;
+#if AZ_DEBUG_BUILD // Clearing the original pointers isn't necessary, but is good for debugging
+            other.m_begin = nullptr;
+            other.m_end = nullptr;
+#endif
+            return *this;
+        }
+
+        size_type size() const { return m_end - m_begin; }
+
+        bool empty() const { return m_end == m_begin; }
+
+        pointer data() { return m_begin; }
+        const_pointer data() const { return m_begin; }
+
+        const_reference operator[](size_type index) const 
+        {
+            AZ_Assert(index < size(), "index value is out of range");
+            return m_begin[index]; 
+        }
+
+        void erase() { m_begin = m_end = nullptr; }
+
+        iterator         begin() const { return m_begin; }
+        iterator         end() const { return m_end; }
+        const_iterator   cbegin() const { return m_begin; }
+        const_iterator   cend() const { return m_end; }
+        reverse_iterator rbegin() const { return reverse_iterator(m_end); }
+        reverse_iterator rend() const { return reverse_iterator(m_begin); }
+        const_reverse_iterator crbegin() const { return const_reverse_iterator(cend()); }
+        const_reverse_iterator crend() const { return const_reverse_iterator(cbegin()); }
+
+        friend bool operator==(span lhs, span rhs)
+        {
+            return lhs.m_begin == rhs.m_begin && lhs.m_end == rhs.m_end;
+        }
+
+        friend bool operator!=(span lhs, span rhs) { return !(lhs == rhs); }
+        friend bool operator< (span lhs, span rhs) { return lhs.m_begin < rhs.m_begin || lhs.m_begin == rhs.m_begin && lhs.m_end < rhs.m_end; }
+        friend bool operator> (span lhs, span rhs) { return lhs.m_begin > rhs.m_begin || lhs.m_begin == rhs.m_begin && lhs.m_end > rhs.m_end; }
+        friend bool operator<=(span lhs, span rhs) { return lhs == rhs || lhs < rhs; }
+        friend bool operator>=(span lhs, span rhs) { return lhs == rhs || lhs > rhs; }
+
+    private:
+        pointer m_begin;
+        pointer m_end;
+    };
```

Comment:
When I gave a try at implementing this a few days ago, I also needed the following template deduction rules.  I'm not a templating wizard though, so I can't speak to whether or not I got them right.  
```
    template<class Container>
    array_view(const Container&) -> array_view<const typename Container::value_type>;

    template<class Container>
    array_view(Container&) -> array_view<typename Container::value_type>;

    template<class Element>
    array_view(typename array_view<Element>::const_pointer, typename array_view<Element>::size_type)
        -> array_view<const typename array_view<Element>::value_type>;

    template<class Element>
    array_view(typename array_view<Element>::pointer, typename array_view<Element>::size_type)
        -> array_view<typename array_view<Element>::value_type>;

    template<class Element>
    array_view(typename array_view<Element>::const_pointer, typename array_view<Element>::const_pointer)
        -> array_view<const typename array_view<Element>::value_type>;

    template<class Element>
    array_view(typename array_view<Element>::pointer, typename array_view<Element>::pointer)
        -> array_view<typename array_view<Element>::value_type>;
```

Reply:
Actually I'll update the span class to properly work use the correct concepts and add the proper deduction guide after SJ gets this in.